dsgrid/config/registration_models.py

```diff
@@ -0,0 +1,129 @@
+"""Contains data models to control registration of test projects and datasets."""
+
+from collections.abc import Iterable
+from pathlib import Path
+from typing import Any, Optional
+
+from pydantic import Field, model_validator
+from typing_extensions import Annotated
+
+from dsgrid.data_models import DSGBaseModel
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.utils.files import load_data
+
+
+class DatasetRegistrationModel(DSGBaseModel):
+    """Defines a dataset to be registered and optionally submitted to a project."""
+
+    dataset_id: Annotated[str, Field(description="Dataset ID")]
+    dataset_path: Annotated[
+        Path, Field(description="Directory containing load_data/load_data_lookup.parquet")
+    ]
+    config_file: Annotated[Path, Field(description="Path to dataset.json5")]
+    dimension_mapping_file: Annotated[
+        Optional[Path],
+        Field(
+            description="Path to file containing mappings of dataset-to-project dimensions",
+            default=None,
+        ),
+    ]
+    dimension_mapping_references_file: Annotated[
+        Optional[Path],
+        Field(
+            description="Path to file containing references to mappings of dataset-to-project dimensions",
+            default=None,
+        ),
+    ]
+    replace_dimension_names_with_ids: Annotated[
+        bool,
+        Field(
+            description="Replace the dimension entries with IDs of dimensions in the database "
+            "with matching names. Typically only useful for tests.",
+            default=False,
+        ),
+    ]
+    replace_dimension_mapping_names_with_ids: Annotated[
+        bool,
+        Field(
+            description="Replace the dimension mapping entries with IDs of dimension mappings "
+            "in the database with matching names. Typically only useful for tests.",
+            default=False,
+        ),
+    ]
+    register_dataset: Annotated[
+        bool, Field(description="If True, register the dataset", default=True)
+    ]
+    submit_to_project: Annotated[
+        bool, Field(description="If True, submit the dataset to the project", default=True)
+    ]
```

Comment:
Which project?

Reply:
I see--this structure goes below the project structure.

But ... if the project_id were supplied here, the bulk register data structure could be flatter, i.e., it could list both projects and datasets, with datasets at the same level as, rather than under, individual projects.

Reply:
Sure, we could redesign the structure as part of making this functionality public. We have a few of these files floating around, like https://github.com/dsgrid/dsgrid/blob/main/tests/data/standard_scenarios_registration.json5. It's been this way a long time.

What's the best way to present these concepts to the user? I typically think of projects containing datasets while understanding that dsgrid treats them separately. We haven't had a case yet where two projects share datasets, though I know it's likely.

If I were to make this change, I would do something like this so that it lines with dsgrid completely:
```
{
  projects: [{project_id: "project1", ...}],
  datasets: [{dataset_id: "dataset1", ...}],
  dataset_submissions_to_projects: [
    {
       dataset_id: "dataset1",
       project_id: "project1",
       dimension_mapping_file: "file_path_project1.json5",
       autogen-reverse-supplemental-mappings: false,
    },
    {
       dataset_id: "dataset1",
       project_id: "project2",
       dimension_mapping_file: "file_path_project2.json5"
       autogen-reverse-supplemental-mappings: false
    },
  ],
}
```

Reply:
I would add that the main use case for this feature is to register a project and all of its datasets. Another option is to remove this list of project objects from this data structure. The project fields could be at the root level. Could we use it to bulk register many projects and datasets? Sure.

Reply:
This conversation got cut short by a fire alarm, but I think the decision was to move forward with the above, since it is future-proofed.